### PR TITLE
add logic to switch away from bad hosts

### PR DESCRIPTION
**Pull request by DavidVorick**
I do not have tests yet for the code here, but this contains the fixes that will allow renters to automatically avoid hosts which no longer have attractive or competitive scores.

I am about to connect the host reliability metrics that @ChrisSchinnerl added in a recent pull request to the scoring system, which means that the contractor will also eventually move past any contracts that are consistently failing during uploads or downloads.

### CODE REVIEW

File: modules/renter/contractor/contractor_test.go [hunk 1]
```diff
@@ -345,7 +351,10 @@ func TestIntegrationSetAllowance(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	if len(c.contracts) != 2 {
+	c.mu.Lock()
+	clen := len(c.contracts)
+	c.mu.Unlock()
+	if clen != 2 {
 		t.Fatal("expected 2 contracts, got", len(c.contracts))
```
Comment (johnathanhowell) [inline modules/renter/contractor/contractor_test.go]: shouldn't this be `clen`?

File: modules/renter.go [hunk 2]
```diff
@@ -224,6 +224,11 @@ type RenterContract struct {
 	ContractFee types.Currency `json:"contractfee"`
 	TxnFee      types.Currency `json:"txnfee"`
 	SiafundFee  types.Currency `json:"siafundfee"`
+
+	// GoodForUpload indicates whether the contract should be used to upload new
+	// data or not.
+	GoodForUpload bool
```
Comment (tlightsky) [inline modules/renter.go]: gooood

File: modules/renter/contractor/contracts.go [hunk 3]
```diff
@@ -356,12 +482,14 @@ func (c *Contractor) threadedContractMaintenance() {
...
 	// contracts.
 	for _, host := range hosts {
-		// Attempt forming a contract with htis host.
+		// Attempt forming a contract with this host.
 		newContract, err := c.managedNewContract(host, numSectors, endHeight)
 		if err != nil {
 			c.log.Printf("Attempted to form a contract with %v, but if failed: %v\n", host.NetAddress, err)
 			continue
 		}
+		// newContract.GoodForUpload = true
```
Comment (johnathanhowell) [inline modules/renter/contractor/contracts.go]: should this be commented out??

File: modules/renter/contractor/contracts.go [hunk 4]
```diff
@@ -83,6 +83,104 @@ func (c *Contractor) contractEndHeight() types.BlockHeight {
...
+	c.mu.RLock()
+	hostCount := int(c.allowance.Hosts)
+	c.mu.RUnlock()
+	hosts := c.hdb.RandomHosts(hostCount+minScoreHostBuffer, nil)
+
+	// Find the lowest score of this batch of hosts.
+	if len(hosts) <= 0 {
+		return
+	}
+	lowestScore := c.hdb.ScoreBreakdown(hosts[0]).Score
```
Comment (johnathanhowell) [inline modules/renter/contractor/contracts.go]: This nondeterministically causes an `index out of range` panic on my machine

File: modules/renter/contractor/contracts.go [hunk 5]
```diff
@@ -83,6 +84,104 @@ func (c *Contractor) contractEndHeight() types.BlockHeight {
...
+	c.mu.RLock()
+	hostCount := int(c.allowance.Hosts)
+	c.mu.RUnlock()
+	if hostCount <= 0 {
+		return
+	}
+	hosts := c.hdb.RandomHosts(hostCount+minScoreHostBuffer, nil)
+
+	// Find the lowest score of this batch of hosts.
+	lowestScore := c.hdb.ScoreBreakdown(hosts[0]).Score
```
Comment (johnathanhowell) [inline modules/renter/contractor/contracts.go]: This nondeterministically causes a `index out of range` panic on my machine. I think we need a `if len(hosts)` branch before this line

**Comment by DavidVorick** [comment]
T.T I just want the build to pass T.T

File: modules/renter/contractor/contracts.go [hunk 6]
```diff
@@ -83,6 +84,119 @@ func (c *Contractor) contractEndHeight() types.BlockHeight {
...
+	}
+	hosts := c.hdb.RandomHosts(hostCount+minScoreHostBuffer, nil)
+	if len(hosts) <= 0 {
+		return
+	}
+
+	// Find the lowest score of this batch of hosts.
+	lowestScore := c.hdb.ScoreBreakdown(hosts[0]).Score
+	for i := 1; i < len(hosts); i++ {
+		score := c.hdb.ScoreBreakdown(hosts[i]).Score
```
Comment (lukechampine) [inline modules/renter/contractor/contracts.go]: `for _, h := range hosts[1:]` also works

File: modules/renter/contractor/contracts.go [hunk 7]
```diff
@@ -83,6 +84,119 @@ func (c *Contractor) contractEndHeight() types.BlockHeight {
...
+		// Contract has no utility if the score is poor.
+		if c.hdb.ScoreBreakdown(host).Score.Cmp(minScore) < 0 {
+			contracts[i].GoodForUpload = false
+			contracts[i].GoodForRenew = false
+			continue
+		}
+		// Contract has no utility if the host is offline.
+		c.mu.Lock()
+		offline := c.isOffline(contracts[i].ID)
+		c.mu.Unlock()
```
Comment (lukechampine) [inline modules/renter/contractor/contracts.go]: why does this require a write-lock?

Comment (DavidVorick) [inline modules/renter/contractor/contracts.go]: convention mandates that any call to an unexported function elsewhere in the module requires a writelock.

File: modules/renter/contractor/contracts.go (hunk 6 above)
Comment (DavidVorick) [inline modules/renter/contractor/contracts.go]: less memory efficient.

File: modules/renter/contractor/contracts.go (hunk 7 above)
Comment (lukechampine) [inline modules/renter/contractor/contracts.go]: why?

File: api/renter.go [hunk 8]
```diff
@@ -278,6 +284,9 @@ func (api *API) renterContractsHandler(w http.ResponseWriter, _ *http.Request, _
 			StorageSpending:  c.StorageSpending,
 			TotalCost:        c.TotalCost,
 			UploadSpending:   c.UploadSpending,
+
+			GoodForRenew:  c.GoodForRenew,
+			GoodForUpload: c.GoodForUpload,
```
Comment (DavidVorick) [inline api/renter.go]: ah whoops, this was for testing

**Comment by DavidVorick** [comment]
I feel like I can't even be mad at this point. 25 failed commits in a row.

A lot of the struggle was because the existing code simply wasn't handling the renew process correctly. Doubling up on hosts, failing to repair dead contracts, etc etc. Basically I wrote a test that dug a lot deeper than any other test we have, and it exposed a lot of issues.

**Comment by DavidVorick** [comment]
Download code is still loose, but this is ready to merge

**Pull request merged**